cli/pcluster/config/config_patch.py

```diff
@@ -0,0 +1,256 @@
+        self.target_config = copy.deepcopy(target_config)
+        self.changes = []
+        self._compare()
+
+    @property
+    def stack_name(self):
+        """Get the name of the stack this patch is referred to."""
+        return (
+            utils.get_stack_name(self.base_config.cluster_name) if hasattr(self.base_config, "cluster_name") else None
+        )
+
+    def _compare(self):
+        """
+        Compare target with base configuration.
+
+        All detected changes are added to the internal changes list, ready to be checked  through the public check()
+        method.
+        """
+        # Remove ignored sections
+        self._remove_ignored_sections(self.base_config)
+        self._remove_ignored_sections(self.target_config)
+
+        # First, compare all sections from target vs base config and mark visited base sections.
+        for section_key in sorted(self.target_config.get_section_keys()):
+            for section_label in sorted(self.target_config.get_sections(section_key).keys()):
+                target_section = self.target_config.get_section(section_key, section_label)
+                base_section = self._get_config_section(self.base_config, target_section)
+                base_section.visited = True
+                self._compare_section(base_section, target_section)
+
+        # Then, compare all non visited base sections vs target config.
+        for section_key in sorted(self.base_config.get_section_keys()):
+            for section_label in sorted(self.base_config.get_sections(section_key).keys()):
+                base_section = self.base_config.get_section(section_key, section_label)
+                if not hasattr(base_section, "visited"):
+                    target_section = self._get_config_section(self.target_config, base_section)
+                    self._compare_section(base_section, target_section)
+
+    def _compare_section(self, base_section, target_section):
+        """
+        Compare the provided base and target sections and append the detected changes to the internal changes list.
+
+        :param base_section: The section in the base configuration
+        :param target_section: The corresponding section in the target configuration
+        """
+        for _, param in target_section.params.items():
+            base_value = base_section.get_param_value(param.key)
+            target_value = param.value
+
+            if not self._compare_params(base_section.key, param.key, base_value, target_value):
+                self.changes.append(
+                    Change(
+                        target_section.key,
+                        target_section.label,
+                        param.key,
+                        base_value,
+                        param.value,
+                        param.get_update_policy(),
+                    )
+                )
+
+    def _compare_params(self, section_key, param_key, base_value, target_value):
```

Comment:
As discussed let's move the comparator method in the `Param` object and add a specific case for the EBSSettingsParam one (see first comment in this file).

In this way instead of having a PARAM_COMPARATORS dictionary here, we can rely on the Param type to find the right comparator to use.

and we can eventually define a compare method for the Section object too, which return the Change object (see my previous comment)

Reply:
outdated